File: ros2_workspace/src/roarm_ws/src/roarm_main/roarm_moveit_servo/scripts/joystickcontrol.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Joy
from geometry_msgs.msg import TwistStamped
from control_msgs.msg import JointJog
from std_msgs.msg import Float32
from roarm_msgs.srv import ServoCommandType


class JoystickServo(Node):
    def __init__(self):
        super().__init__("joystick_servo")

        # ---------------------------
        # Parameters
        # ---------------------------
        self.declare_parameter("model", "roarm_m2")
        self.declare_parameter("enable_twist", True)
        self.declare_parameter("enable_joint", True)
        self.declare_parameter("enable_gripper", True)
        self.declare_parameter("twist_scale_lin", 0.3)
        self.declare_parameter("twist_scale_ang", 0.6)
        self.declare_parameter("joint_vel_scale", 1.0)
        self.declare_parameter("gripper_step", 0.02)

        model = self.get_parameter("model").value
        self.enable_twist = self.get_parameter("enable_twist").value
        self.enable_joint = self.get_parameter("enable_joint").value
        self.enable_gripper = self.get_parameter("enable_gripper").value
        self.twist_scale_lin = self.get_parameter("twist_scale_lin").value
        self.twist_scale_ang = self.get_parameter("twist_scale_ang").value
        self.joint_vel_scale = self.get_parameter("joint_vel_scale").value
        self.gripper_step = self.get_parameter("gripper_step").value

        # Set joint names based on model
        if model == "roarm_m2":
            self.joint_names = [
                "base_link_to_link1",
                "link1_to_link2",
                "link2_to_link3"
            ]
        elif model == "roarm_m3":
            self.joint_names = [
                "base_link_to_link1",
                "link1_to_link2",
                "link2_to_link3",
                "link3_to_link4",
                "link4_to_link5"
            ]
        else:
            self.get_logger().warn(f"Unknown model '{model}', defaulting to M3")
            self.joint_names = [
                "base_link_to_link1",
                "link1_to_link2",
                "link2_to_link3",
                "link3_to_link4",
                "link4_to_link5"
            ]

        self.get_logger().info(f"Using joint names: {self.joint_names}")

        # ---------------------------
        # Publishers
        # ---------------------------
        self.twist_pub = self.create_publisher(TwistStamped, "/servo_node/delta_twist_cmds", 10)
        self.joint_pub = self.create_publisher(JointJog, "/servo_node/delta_joint_cmds", 10)
        self.gripper_pub = self.create_publisher(Float32, "/gripper_cmd", 10)

        # ---------------------------
        # Service client for mode switching
        # ---------------------------
        self.switch_srv = self.create_client(ServoCommandType, "/servo_node/switch_command_type")

        # ---------------------------
        # Joystick subscription
        # ---------------------------
        self.joy_sub = self.create_subscription(Joy, "/joy", self.joy_callback, 10)

        # ---------------------------
        # Internal state
        # ---------------------------
        self.current_frame = "base_link"
        self.gripper_pos = 0.5

        self.get_logger().info("PS4 joystick servo control started.")

    # ---------------------------
    # Mode switching
    # ---------------------------
    def switch_mode(self, mode_value: int):
        if not self.switch_srv.wait_for_service(timeout_sec=0.1):
            self.get_logger().warn("Servo switch service not available")
            return
        req = ServoCommandType.Request()
        req.command_type = mode_value
        self.switch_srv.call_async(req)

    # ---------------------------
    # Joystick callback
    # ---------------------------
    def joy_callback(self, msg: Joy):
        # ---- Buttons ----
        cross = msg.buttons[0]       # X — joint mode
        circle = msg.buttons[1]      # O — twist mode
        square = msg.buttons[2]      # □ — close gripper
        triangle = msg.buttons[3]    # △ — open gripper
        L1 = msg.buttons[4]
        R1 = msg.buttons[5]
        share = msg.buttons[8]       # frame = hand_tcp
        options = msg.buttons[9]     # frame = base_link

        # ---- Axes ----
        LX = msg.axes[0]
        LY = msg.axes[1]
        RX = msg.axes[2]
        RY = msg.axes[3]

        # ---------------------------
        # Switch modes
        # ---------------------------
        if circle:
            self.switch_mode(ServoCommandType.Request.TWIST)
        if cross:
            self.switch_mode(ServoCommandType.Request.JOINT_JOG)

        # ---------------------------
        # Switch command frame
        # ---------------------------
        if share:
            self.current_frame = "hand_tcp"
        if options:
            self.current_frame = "base_link"

        # ---------------------------
        # Twist commands
        # ---------------------------
        if self.enable_twist:
            twist = TwistStamped()
            twist.header.stamp = self.get_clock().now().to_msg()
            twist.header.frame_id = self.current_frame

            twist.twist.linear.x = LY * self.twist_scale_lin
            twist.twist.linear.y = LX * self.twist_scale_lin
            twist.twist.angular.z = RX * self.twist_scale_ang
            twist.twist.linear.z = 0.0
            if L1:
                twist.twist.linear.z = 0.5
            elif R1:
                twist.twist.linear.z = -0.5

            self.twist_pub.publish(twist)

        # ---------------------------
        # Joint jog commands
        # ---------------------------
        if self.enable_joint:
            joint_msg = JointJog()
            joint_msg.header.stamp = self.get_clock().now().to_msg()
            joint_msg.header.frame_id = self.current_frame
            joint_msg.joint_names = self.joint_names
            joint_msg.velocities = [0.0] * len(self.joint_names)

            if len(self.joint_names) > 0:
                joint_msg.velocities[0] = RX * self.joint_vel_scale
            if len(self.joint_names) > 1:
                joint_msg.velocities[1] = RY * self.joint_vel_scale

            self.joint_pub.publish(joint_msg)

        # ---------------------------
        # Gripper commands
        # ---------------------------
        if self.enable_gripper:
            if triangle:  # open
                self.gripper_pos += self.gripper_step
            if square:    # close
                self.gripper_pos -= self.gripper_step

            self.gripper_pos = max(0.0, min(1.5, self.gripper_pos))
            gr_msg = Float32()
            gr_msg.data = self.gripper_pos
            self.gripper_pub.publish(gr_msg)


def main(args=None):
    rclpy.init(args=args)
    node = JoystickServo()
    rclpy.spin(node)
    node.destr


if __name__ == "__main__":
    main()

File: ros2_workspace/src/roarm_ws/src/roarm_main/roarm_moveit_servo/src/joystickcontrol.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joy.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <control_msgs/msg/joint_jog.hpp>
#include <std_msgs/msg/float32.hpp>
#include <roarm_msgs/srv/servo_command_type.hpp>

class JoystickServo : public rclcpp::Node
{
public:
  JoystickServo() : Node("joystick_control"), joint_vel_cmd_(1.0), gripper_value_(0.5)
  {
    // Publishers
    twist_pub_ = create_publisher<geometry_msgs::msg::TwistStamped>(
        "/servo_node/delta_twist_cmds", 10);

    joint_pub_ = create_publisher<control_msgs::msg::JointJog>(
        "/servo_node/delta_joint_cmds", 10);

    gripper_pub_ = create_publisher<std_msgs::msg::Float32>(
        "/gripper_cmd", 10);

    // Mode switch service
    switch_input_ = create_client<roarm_msgs::srv::ServoCommandType>(
        "/servo_node/switch_command_type");

    joy_sub_ = create_subscription<sensor_msgs::msg::Joy>(
        "/joy", 10,
        std::bind(&JoystickServo::joy_callback, this, std::placeholders::_1));

    RCLCPP_INFO(get_logger(), "PS4 Joystick Servo Control Started");
  }

private:
  // ---- Callback handling joystick input ----
  void joy_callback(const sensor_msgs::msg::Joy::SharedPtr msg)
  {
    // ------------- Buttons ---------------
    bool btn_cross = msg->buttons[0];
    bool btn_circle = msg->buttons[1];
    bool btn_square = msg->buttons[2];
    bool btn_triangle = msg->buttons[3];

    bool btn_L1 = msg->buttons[4];
    bool btn_R1 = msg->buttons[5];

    bool btn_share = msg->buttons[8];
    bool btn_options = msg->buttons[9];

    // ------------- Axes (sticks) ----------
    double LX = msg->axes[0];
    double LY = msg->axes[1];
    double RX = msg->axes[2];
    double RY = msg->axes[3];
    double L2 = msg->axes[4];   // usually analog
    double R2 = msg->axes[5];

    // Example mapping:
    // Left stick  -> linear motion
    // Right stick -> rotation
    // L1+R1 -> Z axis up/down
    // ◯ → TWIST mode
    // ❌ → JOINT mode
    // SHARE → EE frame, OPTIONS → Base frame

    // ----------------- Switch Mode -----------------
    if (btn_circle)
      switch_mode(roarm_msgs::srv::ServoCommandType::Request::TWIST);

    if (btn_cross)
      switch_mode(roarm_msgs::srv::ServoCommandType::Request::JOINT_JOG);

    // ----------------- Frame Switch -----------------
    if (btn_share)
      current_frame_ = "hand_tcp";

    if (btn_options)
      current_frame_ = "base_link";

    // ----------------- Twist Command -----------------
    auto twist = geometry_msgs::msg::TwistStamped();
    twist.header.stamp = now();
    twist.header.frame_id = current_frame_;

    twist.twist.linear.x = LY * 0.5;
    twist.twist.linear.y = LX * 0.5;
    twist.twist.angular.z = RX * 1.0;

    // L1 / R1 for Z movement
    if (btn_L1)
      twist.twist.linear.z = 0.5;
    else if (btn_R1)
      twist.twist.linear.z = -0.5;

    twist_pub_->publish(twist);

    // ----------------- Gripper -----------------
    if (btn_triangle)     // open
      gripper_value_ += 0.01;

    if (btn_square)       // close
      gripper_value_ -= 0.01;

    gripper_value_ = std::clamp(gripper_value_, 0.0, 1.5);

    auto g = std_msgs::msg::Float32();
    g.data = gripper_value_;
    gripper_pub_->publish(g);
  }

  // --------- Service to switch input type ----------
  void switch_mode(int mode)
  {
    if (!switch_input_->wait_for_service(std::chrono::milliseconds(10)))
      return;

    auto req = std::make_shared<roarm_msgs::srv::ServoCommandType::Request>();
    req->command_type = mode;
    switch_input_->async_send_request(req);
  }

  // -------- Members --------
  double joint_vel_cmd_;
  double gripper_value_;
  std::string current_frame_ = "base_link";

  rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr joy_sub_;
  rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr twist_pub_;
  rclcpp::Publisher<control_msgs::msg::JointJog>::SharedPtr joint_pub_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr gripper_pub_;
  rclcpp::Client<roarm_msgs::srv::ServoCommandType>::SharedPtr switch_input_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<JoystickServo>());
  rclcpp::shutdown();
  return 0;
}
